src/iter.rs

```diff
@@ -20,62 +21,80 @@ pub struct MiniconfIter<M, S, const L: usize, P> {
+        Self {
+            count: None,
+            separator,
+            state: [0; L],
+            m: PhantomData,
+            p: PhantomData,
+        }
     }
 }
 
-impl<M: SerDe<S>, S, const L: usize, P: Write + Default> Iterator for MiniconfIter<M, S, L, P> {
+impl<'a, M, const L: usize, P> Iterator for PathIter<'a, M, L, P>
+where
+    M: Miniconf + ?Sized,
+    P: Write + Default,
+{
     type Item = P;
 
     fn next(&mut self) -> Option<Self::Item> {
         let mut path = Self::Item::default();
 
         loop {
-            match M::next_path(&self.state, 0, &mut path, M::SEPARATOR) {
-                Ok(depth) => {
-                    self.count = self.count.map(|c| c - 1);
-                    self.state[depth] += 1;
-                    return Some(path);
-                }
-                Err(IterError::Next(0)) => {
+            return match M::path(&mut self.state.iter().copied(), &mut path, self.separator) {
+                // Not having consumed any name/index, the only possible case here is a bare option.
+                // And that can not return `NotFound`.
+                Err(Error::NotFound(0)) => unreachable!(),
+                // Iteration done
+                Err(Error::NotFound(1)) => {
                     debug_assert_eq!(self.count.unwrap_or_default(), 0);
-                    return None;
+                    None
                 }
-                Err(IterError::Next(depth)) => {
+                // Node not found at depth: reset current index, increment parent index, then retry
+                Err(Error::NotFound(depth)) => {
                     path = Self::Item::default();
-                    self.state[depth] = 0;
-                    self.state[depth - 1] += 1;
+                    self.state[depth - 1] = 0;
+                    self.state[depth - 2] += 1;
+                    continue;
+                }
+                // Iteration done for a bare Option
+                Ok(Ok::Leaf(0)) if self.count == Some(0) => None,
+                // Root is `Leaf`: bare Option.
+                // Since there is no way to end iteration by triggering `NotFound` on a bare Option,
+                // we force the count to Some(0) and trigger on that (see above).
+                Ok(Ok::Leaf(0)) => {
+                    debug_assert_eq!(self.count.unwrap_or(1), 1);
+                    self.count = Some(0);
+                    Some(path)
                 }
-                e => {
-                    e.unwrap();
+                // Non-root leaf: advance at current depth
+                Ok(Ok::Leaf(depth)) => {
+                    self.count = self.count.map(|c| c - 1);
+                    self.state[depth - 1] += 1;
+                    Some(path)
                 }
-            }
+                // If we end at a leaf node, the state array is too small.
+                Ok(Ok::Internal(_depth)) => panic!("State too small"),
+                Err(e) => panic!("{e:?}"),
```

Comment:
```suggestion
                Err(e) => panic!("Miniconf path iteration encountered an internal error: {e:?}"),
```

Reply:
Hmm. That doesn't help me. The `Miniconf path iteration` context is better read from the backtrace and the panic message and all `panics` should be internal errors. A panic due to an non-internal error would be an obvious design error.
I did think about this and really didn't see what meaningful information I could add in the message.

Reply:
My main point is for things where the panic trace information is lost (i.e. production). A good example is `booster`'s `service` command. Literally all we have is the serialized `Panic` contents - if we just do a debug print of the error, we don't even know that it relates to `Miniconf`

Reply:
Ack. I added a bit of context to both messages.